### PR TITLE
Read Cabo Verde's BUR1

### PR DESCRIPTION
**Pull request by crdanielbusch**
## Description

Read Cabo Verde BUR1

## Checklist

Please confirm that this pull request has done the following:

- [ ] Tests added
- [ ] Documentation added (where applicable)
- [ ] Changelog item added to `changelog/`

## Open questions

- Are "CO2 emissions from Biomass" and "CO2 emissions from using manure as energy" the same category, namely `M.BIO`? Found on page 86
- Does downscaling make sense here? For example from `1` to `1.A`
- More generally on category aggregation: Should we include categories that are not in the report in aggregation step for the sake of comprehensiveness? For example, I can add the category `5` in the aggregation `"0" : {"sources" : ["1", "2", "3", "4", "5"]}`,  although I already know that there is no data on category `5`.

### CODE REVIEW

File: src/unfccc_ghg_data/unfccc_reader/Cabo_Verde/__init__.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+    #print available datasets
+    print("="*15 + " Available datasets " + "="*15)
+    get_country_datasets(country, True)
+
+You can also obtain this information running
+
+.. code-block:: bash
+
+    poetry run doit country_info country=BDI
```
Comment (JGuetschow) [inline src/unfccc_ghg_data/unfccc_reader/Cabo_Verde/__init__.py:26]:
```suggestion
    poetry run doit country_info country=CPV
```

File: src/unfccc_ghg_data/unfccc_reader/Cabo_Verde/config_cpv_bur1.py [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+}
+
+filter_remove = {
+    "f_memo": {"category": "MEMO"},
+    # They are all NaN and don't match a pre-defined entity
+    "f_fluor": {"entity": "Other fluorinated products"},
+}
+
+meta_data = {
+    "references": "https://unfccc.int/sites/default/files/resource/BUR_EN_Digital.pdf",  # TODO check other sources
```
Comment (JGuetschow) [inline src/unfccc_ghg_data/unfccc_reader/Cabo_Verde/config_cpv_bur1.py:65]: I usually link the overview page for the submission, not the pdf. https://unfccc.int/documents/638907

File: src/unfccc_ghg_data/unfccc_reader/Cabo_Verde/read_CPV_BUR1_from_pdf.py [hunk 3]
```diff
@@ -0,0 +1,309 @@
+"""
+Read Burundi's BUR1 from pdf
```
Comment (JGuetschow) [inline src/unfccc_ghg_data/unfccc_reader/Cabo_Verde/read_CPV_BUR1_from_pdf.py:2]: Cabo Verde

File: docs/source/api/unfccc_ghg_data.helper.functions.rst
Comment (JGuetschow) [inline docs/source/api/unfccc_ghg_data.helper.functions.rst:1]: I've now also regenerated the docs in main. If you pull main into the branch these file should not be part of the PR any more.

File: docs/source/api/unfccc_ghg_data.unfccc_reader.Bangladesh.config_bgd_bur1.rst
Comment (JGuetschow) [inline docs/source/api/unfccc_ghg_data.unfccc_reader.Bangladesh.config_bgd_bur1.rst:1]: Maybe we should merge Bangladesh first so the docs don't appear before the actual code

File: src/unfccc_ghg_data/unfccc_reader/Cabo_Verde/config_cpv_bur1.py
Comment (JGuetschow) [inline src/unfccc_ghg_data/unfccc_reader/Cabo_Verde/config_cpv_bur1.py:1]: Some sectors are missing but subsectors are available (e.g. 2.A, CH4). Please create them even if they are 0 as 0 will be used as 0 in primap-hist and nan will be filled from other sources.

Please downscale to primap-hist sectors using the 2019 values. Using a single year for a longer time period has of course a high error, but here a lot of sectors only have one subsector with data, so the downscaling has no error in these cases.

Comment (JGuetschow) [reply src/unfccc_ghg_data/unfccc_reader/Cabo_Verde/config_cpv_bur1.py:1]: For the other sectors (e.g. 3 for N2O) the error is acceptable as we have not much other data.

File: src/unfccc_ghg_data/unfccc_reader/Cabo_Verde/config_cpv_bur1.py [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+    },
+    # We don't have HFCs and PFCs in the report, hence basket_copy is not relevant
+    "basket_copy": {
+        "GWPs_to_add": ["AR4GWP100", "AR5GWP100", "AR6GWP100"],
+        "entities": ["HFCS"],
+        "source_GWP": gwp_to_use,
+    },
+}
+
+gas_baskets = {
```
Comment (JGuetschow) [inline src/unfccc_ghg_data/unfccc_reader/Cabo_Verde/config_cpv_bur1.py:234]: Why are HFCs not included?

File: src/unfccc_ghg_data/unfccc_reader/Cabo_Verde/config_cpv_bur1.py [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+        "skip_rows_start": 2,
+        "cat_codes_manual": {
+            "Energy": "1",
+            "IPPU": "2",
+            "Agriculture": "M.AG",
+            "LULUCF": "M.LULUCF",
+            "Waste": "4",
+        },
+        "header": ["category", "entity", *trend_years],
+        "unit": ["Gg"] * 4 + ["Gg CO2eq"] + ["Gg"] * 9,
```
Comment (JGuetschow) [inline src/unfccc_ghg_data/unfccc_reader/Cabo_Verde/config_cpv_bur1.py:87]: I think f-gases are also in Gg with some (unknown) conversion factor as the values are very small and not in line with the HFCs data. You could use the HFCs to FGASES ratio to get the conversion factor.

**Review by JGuetschow — CHANGES_REQUESTED** [review]
Looks mostly good, but a few small errors and I would add some missing categories and downscaling

File: docs/source/api/unfccc_ghg_data.helper.functions.rst
Comment (crdanielbusch) [reply docs/source/api/unfccc_ghg_data.helper.functions.rst:1]: Looks better now.

File: src/unfccc_ghg_data/unfccc_reader/Cabo_Verde/config_cpv_bur1.py (hunk 5 above)
Comment (crdanielbusch) [reply src/unfccc_ghg_data/unfccc_reader/Cabo_Verde/config_cpv_bur1.py:87]: Ok, so table 6 for the year 2019 says`0,0112 Gg` for F-gases. The main table on page 87 for the year 2019 says `25,095 CO2eq` for HFCS and NO (zero) for Other fluorinated products. F-gases contain HFCs and other fluorinated products. So I should get the coversion factor with 25,095 / 0,0112 = 2240,625. Does that make sense?

File: src/unfccc_ghg_data/unfccc_reader/Cabo_Verde/config_cpv_bur1.py (hunk 4 above)
Comment (crdanielbusch) [reply src/unfccc_ghg_data/unfccc_reader/Cabo_Verde/config_cpv_bur1.py:290]: I probably shouldn't blindly copy this from the last report

File: src/unfccc_ghg_data/unfccc_reader/Cabo_Verde/config_cpv_bur1.py [hunk 6]
```diff
@@ -0,0 +1,367 @@
...
+                "entities": [
+                    "KYOTOGHG (SARGWP100)",
+                    "KYOTOGHG (AR4GWP100)",
+                    "KYOTOGHG (AR5GWP100)",
+                    "KYOTOGHG (AR6GWP100)",
+                ],
+                "dim": f'category ({coords_terminologies["category"]})',
+            },
+            "2_FGASES": {
+                # error of 3.5% for KYOTOGHG (AR6GWP100)
```
Comment (crdanielbusch) [inline src/unfccc_ghg_data/unfccc_reader/Cabo_Verde/config_cpv_bur1.py:303]: What I still don't understand: where does this incostistency come from? It works for all GWPs except for AR6. It also works for HFCS. The value for 2019 should 35.13 as well.
Screenshot 2024-08-01 at 14 40 58

**Comment by crdanielbusch** [comment]
Alright @JGuetschow this is ready for another review

File: src/unfccc_ghg_data/unfccc_reader/Cabo_Verde/config_cpv_bur1.py
Comment (JGuetschow) [reply src/unfccc_ghg_data/unfccc_reader/Cabo_Verde/config_cpv_bur1.py:1]: 4 (Waste) is a primap-hist sector, but the subsectors are not resolved.

File: src/unfccc_ghg_data/unfccc_reader/Cabo_Verde/config_cpv_bur1.py [hunk 7]
```diff
@@ -0,0 +1,367 @@
...
+        "2": {
+            "sources": ["2.A", "2.B", "2.C", "2.D", "2.E", "2.F", "2.G", "2.H"]
+        },  # consistency check
+        "3": {"sources": ["M.AG", "M.LULUCF"]},  # consistency check
+        # "3.A": {"sources": ["3.A.1", "3.A.2"]}, # consistency check
+        "4": {"sources": ["4.A", "4.B", "4.C", "4.D", "4.E"]},  # consistency check
+    },
+    "basket_copy": {
+        "GWPs_to_add": ["AR4GWP100", "AR5GWP100", "AR6GWP100"],
+        "entities": ["HFCS", "FGASES"],
```
Comment (JGuetschow) [inline src/unfccc_ghg_data/unfccc_reader/Cabo_Verde/config_cpv_bur1.py:259]: Only use HFCS here. FGAS basket has to be built later. Converting both with default factors leads to inconsistencies, because the default factors assume a generic mix of FGASES which is usually not what's underlying the repository (and here we have only HFCs, so it's definitely not a general f-gas mix). So if more detailed information is available always only use that and aggregate later.

File: src/unfccc_ghg_data/unfccc_reader/Cabo_Verde/config_cpv_bur1.py (hunk 6 above)
Comment (JGuetschow) [reply src/unfccc_ghg_data/unfccc_reader/Cabo_Verde/config_cpv_bur1.py:303]: In general there is (currently) no need to downscale f-gas data to subsectors of category 2. In PRIMAP-hist f-gases are currently only resolved at the level of category 2. But this will change in the future, so it doesn't hurt to do the downscaling already.
Your discrepancy comes from using default GWP conversion factors for both the HFCs and the F-gas basket (see above)

File: src/unfccc_ghg_data/unfccc_reader/Cabo_Verde/config_cpv_bur1.py (hunk 4 above)
Comment (JGuetschow) [reply src/unfccc_ghg_data/unfccc_reader/Cabo_Verde/config_cpv_bur1.py:358]: happens all the time... A good reason for a function that checks for completeness of the results.

**Review by JGuetschow — CHANGES_REQUESTED** [review]
Looks good except for f-gases / HFCs. I commented on the origin of  the discrepancy and how to overcome it.

**Comment by JGuetschow** [comment]
I've made a few changes on the downscaling and gas basket aggregation to remove inconsistencies. In general it's advisable to only downscale baskets contents to sectors and not also the baskets (if both exist), as the downscaling of the contents uses more information and will usually be inconsistent with the basket downscaling that misses the information how the gases are distributed over the sectors. from my side it's now ready to be merged (I left some comments on commented out section of the downscaling. These could be removed before merging, they are just for information why I deleted the sections.)
